### PR TITLE
feat(locking): Locking module

### PR DESCRIPTION
**Pull request by carlos-r-l-rodrigues**
**What**
- Locking Module to manage concurrency
- Default `in-memory` provider

### CODE REVIEW

File: packages/core/modules-sdk/src/medusa-app.ts [hunk 1]
```diff
@@ -84,7 +84,7 @@ export async function loadModules(args: {
     sharedResourcesConfig,
     migrationOnly = false,
     loaderOnly = false,
-    workerMode = "server" as ModuleBootstrapOptions["workerMode"],
+    workerMode = "shared" as ModuleBootstrapOptions["workerMode"],
```
Comment (olivermrbl) [inline packages/core/modules-sdk/src/medusa-app.ts:87]: q: What happened here?

File: packages/core/utils/src/modules-sdk/definition.ts [hunk 2]
```diff
@@ -24,6 +24,7 @@ export const Modules = {
   FILE: "File",
   NOTIFICATION: "Notification",
   INDEX: "Index",
+  LOCKING: "Locking",
```
Comment (olivermrbl) [inline packages/core/utils/src/modules-sdk/definition.ts]: todo: Should be `locking`
```suggestion
  LOCKING: "locking",
```

File: packages/modules/locking/src/loaders/providers.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      }
+      container.register(LockingDefaultProvider, asValue(provider.id))
+    }
+  }
+
+  if (!hasDefaultProvider) {
+    logger.warn(
+      `No default Locking provider explicit defined, using "${container.resolve(
+        LockingDefaultProvider
+      )}" as default.`
```
Comment (olivermrbl) [inline packages/modules/locking/src/loaders/providers.ts]:
```suggestion
      `No default locking provider defined. Using "${container.resolve(
        LockingDefaultProvider
      )}" as default.`
```

File: packages/modules/locking/src/types/index.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import {
+  IEventBusModuleService,
+  Logger,
+  ModuleProviderExports,
+  ModuleServiceInitializeOptions,
+} from "@medusajs/framework/types"
+
+export type InitializeModuleInjectableDependencies = {
+  logger?: Logger
+  EventBus?: IEventBusModuleService
```
Comment (olivermrbl) [inline packages/modules/locking/src/types/index.ts]:
```suggestion
  eventBus?: IEventBusModuleService
```

File: packages/modules/locking/src/providers/in-memory.ts [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+      if (lock.ownerId === ownerId) {
+        if (expire) {
+          lock.expiration = now + expire * 1000
+          this.locks.set(key, lock)
+        }
+
+        continue
+      }
+
+      if (lock.currentPromise && args?.awaitQueue) {
```
Comment (olivermrbl) [inline packages/modules/locking/src/providers/in-memory.ts:112]: q: can you briefly explain this `awaitQueue`?

**Review by olivermrbl — APPROVED** [review]
LGTM, a few comments. Can you add the module to the large changeset we have?

File: packages/core/modules-sdk/src/medusa-app.ts (hunk 1 above)
Comment (carlos-r-l-rodrigues) [reply packages/core/modules-sdk/src/medusa-app.ts:87]: this is supposed to be "shared" by default.
we have already other places setting this as "shared" if not provided, it is just to align with the real default.

File: packages/modules/locking/src/providers/in-memory.ts (hunk 5 above)
Comment (carlos-r-l-rodrigues) [reply packages/modules/locking/src/providers/in-memory.ts:112]: the methods `execute` and `acquire` use the same mechanism. But the `execute` has to `awaitQueue` to be executed if the lock is acquired by another process, whereas the `acquire` will fail if the lock is not acquired.

Comment (olivermrbl) [reply packages/modules/locking/src/providers/in-memory.ts:112]: Got it thanks!

**Pull request merged**